File: 22.generate-parentheses/go/solution.go
package main

import "fmt"

func generateParenthesis(n int) []string {

	MAXLEN := n * 2
	ans := []string{}
	var backtrack func([]string, string)
	backtrack = func(s []string, comb string) {
		if len(comb) == MAXLEN {
			if len(s) == 0 {
				ans = append(ans, comb)
			}
			return
		}
		backtrack(append(s[:], "("), comb+"(")
		if len(s) > 0 {
			s = s[:len(s)-1]
			backtrack(s, comb+")")
		}
	}

	backtrack([]string{"("}, "(")

	return ans
}

func main() {
	fmt.Println(generateParenthesis(3))
}

File: 22.generate-parentheses/python/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        MAXLEN = n * 2
        answer = []

        def backtrack(stack, combination):
            if len(combination) == MAXLEN:
                if len(stack) == 0:
                    answer.append(combination)
                return
            backtrack(stack + ["("], combination + "(")
            if len(stack) > 0:
                stack.pop()
                backtrack(stack, combination + ")")

        backtrack(["("], "(")

        return answer


print(Solution().generateParenthesis(3))
